### DISCUSSION TITLE
How to handle intermittent connection errors gracefully

### DISCUSSION THREAD

**Discussion by PeterBocan** [nginx/nginx#200] [Q&A]
Hello community, 

we are hosting NGINX as a reverse proxy on AWS ECS w/ Fargate, we use AWS Cloud Map for service discovery). However, we do face the intermittent connection issues, as follows:

```
2024/09/19 09:06:33 [error] 19#19: *31763 connect() failed (111: Connection refused) while connecting to upstream, client: 10.2.2.169, server: _, request: "GET /...? HTTP/1.1", upstream: "http://10.2.3.107:8092/...?", host: "...", referrer: "http://localhost:8080/"
```

I am curious, is there a way to address this issue somehow? Handle it with grace, or add a retry? 

Thank you for your help, 
Peter.

**Comment by barrymw** [comment]
@PeterBocan , What OS is your ECS instance based on?  I.e. is it an EL8 derivative? 

I have seen a similar increase in spurious _failed (111: Connection refused)_ messages when using RHEL8.  Reverting back to RHEL7, the number of these messages is negligible.  When nginx was on RHEL8, the end users never reported any issue with the application itself, so I suspect that nginx is actually handling it 'gracefully', albeit with an additional overhead of re-establishing connectivity to the backend host/s.

**Reply by PeterBocan** [reply]
Hey Mark, we are using Fargate, the docker container is derived from Alpine linux.

**Comment by Maryna-f5** [comment]
This discussion appears to focus on the configuration, troubleshooting or general questions around running NGINX, which is not the intended use for GitHub discussions. 

We encourage you to visit [https://community.nginx.org ](https://community.nginx.org/)and share your discussion there as the NGINX Community Forum is the most appropriate place to engage for these types of questions.

**Comment by Raphasha27** [comment]
Use `proxy_next_upstream` to handle intermittent errors gracefully. It retries failed requests on the next upstream server:

```nginx
upstream backend {
    server backend1:8080;
    server backend2:8080;
}

server {
    location / {
        proxy_pass http://backend;
        
        # Retry on these errors
        proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;
        
        # Don't retry on idempotent-safe methods only (default)
        proxy_next_upstream_tries 3;
        proxy_next_upstream_timeout 10s;
        
        # Wait up to 5s for each upstream
        proxy_connect_timeout 5s;
        proxy_read_timeout 30s;
    }
}
```

For websockets or long-lived connections:

```nginx
location /ws/ {
    proxy_pass http://websocket-backend;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    
    # Retry connection errors (but not after websocket established)
    proxy_next_upstream error timeout invalid_header;
}
```

To avoid retrying on non-idempotent requests (POST, PATCH, DELETE):

```nginx
proxy_next_upstream error timeout;
proxy_next_upstream_timeout 0;  # non-idempotent: don't retry
```

For a single upstream server, `proxy_next_upstream` won't help directly—combine with DNS-based load balancing:

```nginx
resolver 8.8.8.8 valid=10s;

set $backend_upstream "backend.example.com";
proxy_pass http://$backend_upstream;
proxy_next_upstream error timeout;
```

With multiple upstreams, connection errors drop the failing server and retry on the next one transparently.
